File: 1052-grumpy-bookstore-owner/1052-grumpy-bookstore-owner.js
/**
 * @param {number[]} customers
 * @param {number[]} grumpy
 * @param {number} minutes
 * @return {number}
 */
var maxSatisfied = function(customers, grumpy, minutes) {
    let max = 0;
    let issues = [];
    let freebies = 0;
    
    // check which minutes truly count when using its secret technique
    for(let i = 0; i < grumpy.length; i++){
        issues.push(customers[i]*grumpy[i])
        if(grumpy[i] == 0) freebies += customers[i]
    }
    // console.log(issues)
    // start max with the first window
    for(let i = 0; i < minutes; i++){
        max += issues[i]
    }
    // console.log(max)
    let newMax = max
    // use a window to check when it is better to use the secret technique
    for(let i = 0; (i+minutes) < issues.length; i++){
        newMax = newMax - issues[i] + issues[i+minutes];
        if(newMax > max) max = newMax
    }
    return max + freebies
};

File: 1052-grumpy-bookstore-owner/1052-grumpy-bookstore-owner.py
class Solution:
    def maxSatisfied(self, customers: List[int], grumpy: List[int], minutes: int) -> int:
        issue = []
        freebies = 0
        for i in range(len(grumpy)):
            if grumpy[i] == 1:
                issue.append(customers[i])
            else:
                issue.append(0)
                freebies += customers[i]

        m = 0
        
        for i in range(0,minutes):
            m += issue[i]
            
        newMax = m
        
        for i in range(0, len(issue)-minutes):
            newMax = newMax - issue[i] + issue[i+minutes]
            if newMax > m:
                m = newMax
        return m + freebies